tests/sub_group/sub_group_api.cpp

```diff
@@ -0,0 +1,368 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Copyright (c) 2023 The Khronos Group Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+*******************************************************************************/
+
+#include "../common/common.h"
+
+#include <algorithm>
+#include <bitset>
+#include <limits>
+#include <numeric>
+#include <sstream>
+#include <vector>
+
+/** Each of the functions called by the work-items in the below test. */
+enum struct method : int {
+  group_id = 0,
+  local_id,
+  local_range,
+  group_range,
+  max_local_range,
+  group_linear_id,
+  local_linear_id,
+  group_linear_range,
+  local_linear_range,
+  leader,
+  group_group_linear_id,  // part of group, not part of sub-group like above
+  method_count            // defines size, should be last
+};
+
+static constexpr auto method_cnt = to_integral(method::method_count);
+
+struct kernel_name;
+
+/** Return a linear index for a global work-item index and a method. */
+static size_t get_idx(size_t id, method method) {
+  return id * method_cnt + to_integral(method);
+}
+
+/**
+ * Returns the largest possible size that each dimension of a three dimensional
+ * work-group may have such that it is a power of two and
+ * smaller or equal to 1024. */
+static size_t get_work_group_size(const sycl::device& device) {
+#if !(defined(SYCL_CTS_COMPILING_WITH_HIPSYCL) || \
+      defined(SYCL_CTS_COMPILING_WITH_COMPUTECPP))
+  const sycl::id<3> max_work_item_sizes =
+      device.get_info<sycl::info::device::max_work_item_sizes<3>>();
+#else
+  const sycl::id<3> max_work_item_sizes{1, 1, 1};
+  WARN(
+      "Implementation does not define device.get_info<max_work_item_sizes>. "
+      "Using work-group of size (1, 1, 1).");
```

Comment:
Why do we care about testing `sub_group`'s API when `info::device::max_work_item_sizes` is not defined? Because the sub-group may be developed independently of the information descriptors and therefore it would be nice to still be able to run the sub-group test when information descriptors are not implemented. Also, it is relatively little effort to put this workaround in place.